lib/bitcoin/BitcoinProcessor.ts

```diff
@@ -0,0 +1,590 @@
+      const error = new Error(`Broadcast failure [${responseObject.status}]: ${responseData}`);
+      console.error(error);
+      throw error;
+    }
+
+    const response = JSON.parse(responseData);
+
+    return response.success;
+  }
+
+  /**
+   * Will process transactions every interval seconds.
+   * @param interval Number of seconds between each query
+   */
+  private periodicPoll (interval: number = this.pollPeriod) {
+    if (this.pollTimeoutId) {
+      clearTimeout(this.pollTimeoutId);
+    }
+    this.processTransactions(this.lastSeenBlock).then((syncedTo) => {
+      this.lastSeenBlock = syncedTo;
+      this.pollTimeoutId = setTimeout(this.periodicPoll.bind(this), 1000 * interval, interval);
+    }).catch((error) => {
+      console.error(error);
+      throw error;
+    });
+  }
+
+  /**
+   * Processes transactions from startBlock (or genesis) to endBlock (or tip)
+   * @param startBlock The block to begin from
+   * @param endBlock The blockheight to stop on (inclusive)
+   * @returns The block height and hash it processed to
+   */
+  private async processTransactions (startBlock?: IBlockInfo, endBlock?: number): Promise<IBlockInfo> {
+    let beginBlockHeight: number;
+    if (startBlock) {
+      const startValid = await this.verifyBlock(startBlock.height, startBlock.hash);
+      beginBlockHeight = startBlock.height;
+      if (!startValid) {
+        beginBlockHeight = await this.revertBlockchainCache();
+      }
+    } else {
+      beginBlockHeight = this.genesisBlockNumber;
+    }
+    if (endBlock === undefined) {
+      endBlock = await this.getCurrentBlockHeight();
+    }
+
+    console.info(`Processing transactions from ${startBlock} to ${endBlock}`);
+
+    // You can parrallelize this so long as all processBlock's don't throw
+    for (let blockHeight = beginBlockHeight; blockHeight < endBlock; blockHeight++) {
+      await this.processBlock(blockHeight);
+    }
+    const hash = await this.processBlock(endBlock);
+    console.info(`Finished processing blocks ${startBlock} to ${endBlock}`);
+    return {
+      hash,
+      height: endBlock
+    };
+  }
+
+  /**
+   * Begins to revert the blockchain cache until consistent, returns last good height
+   * @returns last valid block height before the fork
+   */
+  private async revertBlockchainCache (): Promise<number> {
+    console.info('Reverting transactions');
+    while (await this.transactionStore.getTransactionsCount() > 0) {
+      const exponentiallySpacedTransactions = await this.transactionStore.getExponentiallySpacedTransactions();
+
+      const firstValidTransaction = await this.firstValidTransaction(exponentiallySpacedTransactions);
+
+      let revertToTransactionNumber: number;
+
+      if (firstValidTransaction) {
+        revertToTransactionNumber = firstValidTransaction.transactionNumber;
+      } else {
+        const lowestHeight = exponentiallySpacedTransactions.reduce((height: number, transaction: ITransaction): number => {

```

Comment:
No need for the reduce call to sort, just get the last transaction from `exponentiallySpacedTransactions`